ports/raspberrypi/common-hal/pulseio/PulseIn.c

```diff
@@ -0,0 +1,243 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021 Dave Putz for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "src/rp2_common/hardware_gpio/include/hardware/gpio.h"
+#include "src/rp2_common/hardware_irq/include/hardware/irq.h"
+
+#include <stdint.h>
+
+#include "py/runtime.h"
+#include "shared-bindings/microcontroller/__init__.h"
+#include "shared-bindings/pulseio/PulseIn.h"
+#include "shared-bindings/microcontroller/Pin.h"
+#include "supervisor/shared/translate.h"
+#include "bindings/rp2pio/StateMachine.h"
+#include "common-hal/pulseio/PulseIn.h"
+
+pulseio_pulsein_obj_t* save_self;
+
+#define NO_PIN 0xff
+volatile bool last_level;
+volatile uint16_t level_count = 0;
+volatile uint16_t result = 0;
+volatile uint16_t buf_index = 0;
+
+uint16_t pulsein_program[] = {
+    0xe03f, //  0: set    x, 31
+    0x4001, //  1: in     pins, 1
+    0x0041, //  2: jmp    x--, 2
+    0x8060, //  3: push   iffull block
+    0xc020, //  4: irq    wait 0
+    0x0000, //  5: jmp    1
```

Comment:
I think this can be simplified greatly to simply `in pins, 1`. IN will auto-push every 32 bits so no need for using x and block when the FIFO is full. Every FIFO has a corresponding interrupt as well so you don't need to manage an interrupt manually either. The final jump isn't needed due to automatic wrapping.

Reply:
Current version uses the StateMachine.c interrupt setup and has simplified the SM program to the single IN instruction.